lib/src/strings.dart

```diff
@@ -67,142 +54,199 @@ String _yamlEncodeDoubleQuoted(String string) {
+    ///   2. If the current line doesn't begin with white-space
+    ///
+    /// Such that we apply `\n` for `foo\nbar` but not `foo\n bar`.
+    if (current.trim().isNotEmpty &&
+        trimmedSplit[index - 1].trim().isNotEmpty &&
+        !current.replaceFirst(indent, '').startsWith(' ')) {
+      updated = lineEnding + updated;
     }
 
-    return _tryYamlEncodePlain(value.value);
-  }
+    /// Apply a `\n` by default.
+    return previous + lineEnding + updated;
+  });
 
-  assertValidScalar(value);
-  return _tryYamlEncodePlain(value);
+  return '>-\n'
+      '$indent$trimmed'
+      '${stripped.replaceAll('\n', lineEnding + indent)}';
 }
 
-/// Returns [value] with the necessary formatting applied in a block context
-/// if possible.
+/// Attempts to encode a [string] as a _YAML literal string_ and apply the
+/// appropriate _chomping indicator_.
 ///
-/// If [value] is a [YamlScalar], we try to respect its [YamlScalar.style]
-/// parameter where possible. Certain cases make this impossible (e.g. a folded
-/// string scalar 'null'), in which case we will produce [value] with default
-/// styling options.
-String yamlEncodeBlockScalar(
-  YamlNode value,
-  int indentation,
-  String lineEnding,
-) {
-  if (value is YamlScalar) {
-    assertValidScalar(value.value);
+/// Returns `null`, if the [string] cannot be encoded as a _YAML literal
+/// string_.
+///
+/// **Examples** of literal strings.
+/// ```yaml
+/// # With the "strip" chomping indicator
+/// key: |-
+///   my literal
+///   string
+///
+/// # Without chomping indicator
+/// key: |
+///   my literal
+///   string
+/// ```
+///
+/// See: https://yaml.org/spec/1.2.2/#812-literal-style
+String? _tryYamlEncodeLiteral(
+    String string, int indentSize, String lineEnding) {
+  if (string.isEmpty || string.trimLeft().length != string.length) return null;
 
-    final val = value.value;
-    if (val is String) {
-      if (_hasUnprintableCharacters(val)) {
-        return _yamlEncodeDoubleQuoted(val);
-      }
+  // A string that starts with space or newline followed by space can't be
+  // encoded in literal mode.
+  if (_hasUnprintableCharacters(string)) return null;
 
-      if (value.style == ScalarStyle.SINGLE_QUOTED) {
-        return _tryYamlEncodeSingleQuoted(val);
-      }
+  // TODO: Are there other strings we can't encode in literal mode?
 
-      // Strings with only white spaces will cause a misparsing
-      if (val.trim().length == val.length && val.isNotEmpty) {
-        if (value.style == ScalarStyle.FOLDED) {
-          return _tryYamlEncodeFolded(val, indentation, lineEnding);
-        }
+  final indent = ' ' * indentSize;
 
-        if (value.style == ScalarStyle.LITERAL) {
-          return _tryYamlEncodeLiteral(val, indentation, lineEnding);
-        }
-      }
-    }
+  /// Simplest block style.
+  /// * https://yaml.org/spec/1.2.2/#812-literal-style
+  return '|-\n$indent${string.replaceAll('\n', lineEnding + indent)}';
```

Comment:
This is better than what we had, so let's accept it.

But I suspect this could cause issues for strings that end in empty lines or whitespace, or combination thereof.